proposals/0047-syscall-and-sysvar-for-last-restart-slot.md

```diff
@@ -0,0 +1,148 @@
+restart is successful, which will add complexity. The best way is to create a
+new syscall and sysvar to get this information during the execution of a
+transaction, which will help us get the last restart slot. We prefer syscall
+over sysvar because the newly created sysvar has to be passed in the instruction
+this takes up transaction accounts space. In the long-term Solana evolution, we
+plan to get rid of syscalls and make everything a sysvar, so we have decided to
+implement both a syscall and a sysvar to address short and long-term evolution.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+The following criteria should be met to choose a restart slot.
+
+* Should be of type unsigned int 64 bits.
+* Should be indicative of when the cluster was restarted
+* Should be monotonically increasing
+* Should be less than equal to the current slot
+* Should be synchronized in the whole cluster
+* Can also be indicative of a slow block
+* The first restart slot should be `0`
+
+In Solana client `hard fork` variable satisfies nearly all the conditions except
+it is not indicative of slow block, it is a good candidate.
+
+The syscall and sysvar will be available in architectures eBPF and SBF. Only the
+sysvar will be available from architectures SBFv2.
+
+### Feature Gate
+
+This feature should be protected by the feature gate this is because this
+feature can create differences in consensus. Solana core developers should
+create a new keypair and use it to enable this feature on the cluster.
+
+Upon activation, the functionality described below immediately becomes
+available.
+
+### Creation of a new sysvar
+
+Similar to the clock or rent sysvar we can create a new sysvar for last restart
+slot. We have to implement a new class to store the required values, which then
+could be used by sysvar during execution.
+
+Sysvar Id: `SysvarLastRestartS1ot1111111111111111111111`
+
+### Creation of a new syscall
+
+The implementation of this syscall is pretty straightforward. The syscall should
+have following signature:
+
+`fn sol_get_last_restart_slot() -> u64`
+
+Murmur3 hash is: `0xb532af38`
+
+### Overview of changes for Solana client
+
+In Solana Labs validator client all the cluster restart-related data is already
+stored in the bank structure. For other validator clients, we have to get the
+last restart slot information and make it accessible to the runtime of the
+program. We have to set this data in invoke context executing the program. Then
+the syscall and sysvar can use invoke context to return the data to the
+executing program.
+
+### Compute budget for the syscall
+
+As this feature involves moving a u64 from the bank to the invoke context and
+returning it when syscall or sysvar is called.
+
+So considering :
+
+```
+/// Cluster averaged compute unit to micro-sec conversion rate
+COMPUTE_UNIT_TO_US_RATIO = 30;
+```
+
+We can set the maximum compute limit for the syscall to:
+
+```
+COMPUTE_UNITS_LIMIT = 2 * COMPUTE_UNIT_TO_US_RATIO
+```
```

Comment:
Yeah these numbers are somewhere in the solana validator code. There are some constants specified for each program.

Reply:
It is possible to define a single fixed CU price for each invocation of the syscall. However, the compute budget is a measure of the overall computational resources that are allocated to a particular process or system. In order to determine the appropriate compute budget for a given system, it is important to take into account various factors such as the complexity of the system, the number of users, and the expected usage patterns.

Therefore, it may be beneficial to have a more nuanced approach to defining the compute budget for the syscall, taking into account these various factors. However, as you mentioned, since this doc is still a draft, it's acceptable to use an arbitrary number for now and revise it later based on further analysis and testing. It's also a good idea to consult the CU defaults to get a baseline understanding of the expected CU consumption for the syscall.